### ISSUE TITLE
How to save a Lightning model that contains a PyTorch model with customized saving function

### ISSUE THREAD

**Issue by junwen-austin** [PyTorchLightning/pytorch-lightning#3096]
#### What is your question?
I'd like to use Lightning to do the training of a PyTorch transformer model. So I wrap the transformer model in a LightningModule. Before training, the model is initialized from a pre-trained model by use of hparams. I use default saving with Lightning. However, the transformer itself has a very specific save/load method. Ideally we would like the transformer model to save using its own method instead of relying on Lightning. My questions are:

* How do I save the Lightning model correctly so that it uses the specific save/load method for the wrapped transformer model?
* How do I load the saved checkpoint correctly so that it will not load the pre-trained model indicated by hparams?

#### Code

```
import transformers
class Transformer(LightningModule):
  def __init__(self, hparams):
    ...

    # Initialize the pytorch model (dependent on an external pre-trained model)
    self.transformer = transformers.from_pretrained(params.transformer_name)

    # note: self.transformer has a method save_pretrained to save it in a directory so ideally we would like it to be saved with its own method instead of default one provided by Lightning

model = Transformer(hparams)
trainer = Trainer(..)
trainer.fit(model)
```

#### What have you tried?
I tried the default saving with Lightning but not sure if this is correct.

#### What's your environment?

 - OS: Linux
 - Packaging pip
 - Version 0.9.0

**Comment by Borda** [comment]
I would use save/load just weights, or any other suggestion @PyTorchLightning/core-contributors?

**Comment by junwen-austin** [comment]
@Borda , thanks for your suggestion. I am aware of this save/load weights. The issue is that the model inside the LightningModule has a customized save/load function: from_pretranied and save_pretrained that takes care of save/load of weights. Ideally, I am looking for something like:


* When Lightning is auto save LightningModule to a checkpoint location:
    * call self.model.save_pretrained(the checkpoint location)
    * save other Lightning stuff (like saving trainer/optimizer state)


* When Lightning is initialize the model from a checkpoint location
    * call self.model.from_pretrained(the checkpoint location)
    * load other lightning stuff (like loading trainer/optimizer state) 
   

In this way, the save/load is passed on from LightningModule(self) to its member (self.module). 

Could core develops chime in and offer recommendation? Thanks a lot!

**Comment by rohitgr7** [comment]
> call self.model.save_pretrained(the checkpoint location)
>  call self.model.from_pretrained(the checkpoint location)

why do you need these methods if you have trained and saved the model and want to load the saved checkpoint?

**Comment by junwen-austin** [comment]
@rohitgr7 The reason I need to use these methods above is that the LightningModule (self) is a wrapper of the pretrained transformer model(self.model). In addition, the self.model has to be loaded from an external location using its method `from_pretrained` and saved using its own method `save_pretrained`. I have copied some code above for your reference:



```
import transformers
class Transformer(LightningModule):
  def __init__(self, hparams):
    ...

    # Initialize the pytorch model (dependent on an external pre-trained model)
    self.transformer = transformers.from_pretrained(hparams.transformer_name)

    # note: self.transformer has a method save_pretrained to save it in a directory so ideally we would like it to be saved with its own method instead of default one provided by Lightning
```

Thanks and looking forward to your comments/suggestions!

**Comment by rohitgr7** [comment]
Ok, `from_pretrained` is ok since you want to load pretrained weights but `save_pretrained` doesn't do anything special apart from saving weights which lightning also does.

**Comment by junwen-austin** [comment]
@rohitgr7 thanks for your comments.

I reloaded the Lightning model saved automatically and found the weighs inside the self.model changed, which led me to think there might be something inside `save_pretrained` that might be different from just saving the weights. Therefore I was hoping that I could customize the saving by myself to respect he `save_pretrained`. 

This could also be an issue of `from_pretrained` inside the `__init__` as in the first run, the weights were loaded from the external data. When there is a need to reload the model from the Lightning, I was not sure which one happens first, the loading from checkpoint file or the loading from pre_trained? 

Really appreciate your insight!!!

**Comment by rohitgr7** [comment]
`save_pretrained` is not doing anything different ([source code](https://github.com/huggingface/transformers/blob/858b7d5873577d7997918439eb7429d424d07dd5/src/transformers/modeling_utils.py#L617-L652)). In your current use-case, when you load the checkpoint using it first loads from `from_pretrained` and then loads the checkpoint weights. So in-short weights loaded from `from_pretrained` is basically getting replaced by the weights from the checkpoint. When you reloaded the LightningModel checkpoint they will definitely be different from the pre-trained weights if you trained the backbone or model in an un-freezed state.

**Comment by junwen-austin** [comment]
@rohitgr7 @awaelchli Thanks for looking into the source code and you are right in this case, there is nothing fancy about the `from_pretrained` in terms of saving the weights. However, for some Transformer models, for example, Albert Transformer, the weights are shared across many layers, and the load/saving functions from Albert Transformer takes advantage of that and the model saved is much smaller than using general save from Pytorch or the checkpoint from Lightning (roughly 14MB vs 140MB). 

This goes back to my point, we might want to respect the customized saving/loading function a model inside a Lightning and provide a way for the model to call its own saving/loading function when Lightning checkpoints.

I'd like to hear your thoughts. Thanks.

**Comment by rohitgr7** [comment]
@junwen-austin which albert version are you referring here??

**Comment by junwen-austin** [comment]
@rohitgr7 the base of Albert, which is about 14MB

I am using AdamW for the optimizer. Is it possible that the additional saving Lightning does (optimizer state and others) becomes as large as more than 100MB? Thanks.

**Comment by rohitgr7** [comment]
I checked Albert `albert-base-v1` and it's around 45mb.

**Comment by junwen-austin** [comment]
@rohitgr7 You are right. Sorry about that. For this case, then it is fine we have 100MB of additional states.

I really do not want to bug you further but do you have a general suggestion as to how to save a LightningModule that wraps a Pytorch model that has some customized saving function so that the Pytorch model has a chance to call it? Thanks.

**Comment by rohitgr7** [comment]
not sure why you want that since both are exactly doing the same thing. But still one solution I can suggest is to override `Trainer.save_checkpoint`. 
```python
class SubTrainer(Trainer):
    def save_checkpoint(self, filepath, weights_only=False):
        if self.is_global_zero:
            dirpath = os.path.split(filepath)[0]
            lightningmodel = self.get_model()
            lightningmodel.transformer.save_pretrained(dirpath)


model = Transformer()
trainer = SubTrainer(**trainer_params)   # with checkpoint callback
trainer.fit(model)
```

if you don't want to use `ModelCheckpoint` and want to save it like after every epoch or something can you can do the same in `training_epoch_end` or `validation_epoch_end` as per your requirement.

[Reference](https://github.com/PyTorchLightning/pytorch-lightning/blob/ee72271d205beb2c012e34425cb16e189cc56c7d/pytorch_lightning/trainer/training_io.py#L273)

**Comment by yukw777** [comment]
@junwen-austin you're trying to save the best model at the end of the training using the method from Hugging Face transformers correct? I encountered a similar problem at work, and my suggestion is to avoid mixing PL's serialization and Hugging Face's serialization. You're just asking for more trouble. More concretely, I suggest you do something like the following:

```python
import transformers
class Transformer(LightningModule):
  def __init__(self, hparams):
    ...

    # Initialize the pytorch model (dependent on an external pre-trained model)
    self.transformer = transformers.from_pretrained(params.transformer_name)

    # note: self.transformer has a method save_pretrained to save it in a directory so ideally we would like it to be saved with its own method instead of default one provided by Lightning

model = Transformer(hparams)
trainer = Trainer(..)
trainer.fit(model)

for i, (path, _) in enumerate(trainer.checkpoint_callback.best_k_models.items()):
    m = Transformer.load_from_checkpoint(path)
    m.transformer.save_pretrained(f'{i}th_best.pt')
```

Yes, you're not calling `save_pretrained()` during training automatically, but you can always load your lightning module from a vanilla PL checkpoint, then call `save_pretrained()` directly.

**Comment by junwen-austin** [comment]
@rohitgr7 @yukw777 Thanks for the suggestions! Very helpful!

Yes, I was to save the best model at the end of the training. What you suggested is also a very good way of extracting the transformer model from Lightning checkpoints.

**Comment by g8a9** [comment]
Hi there. I'm having issues with a similar situation and I think there is something to be fixed in PL's serialization when modules contain HuggingFace's transformers models.

I tried the solution proposed here by @yukw777:

> @junwen-austin you're trying to save the best model at the end of the training using the method from Hugging Face transformers correct? I encountered a similar problem at work, and my suggestion is to avoid mixing PL's serialization and Hugging Face's serialization. You're just asking for more trouble. More concretely, I suggest you do something like the following:
> 
> ```python
> import transformers
> class Transformer(LightningModule):
>   def __init__(self, hparams):
>     ...
> 
>     # Initialize the pytorch model (dependent on an external pre-trained model)
>     self.transformer = transformers.from_pretrained(params.transformer_name)
> 
>     # note: self.transformer has a method save_pretrained to save it in a directory so ideally we would like it to be saved with its own method instead of default one provided by Lightning
> 
> model = Transformer(hparams)
> trainer = Trainer(..)
> trainer.fit(model)
> 
> for i, (path, _) in enumerate(trainer.checkpoint_callback.best_k_models.items()):
>     m = Transformer.load_from_checkpoint(path)
>     m.transformer.save_pretrained(f'{i}th_best.pt')
> ```
> 
> Yes, you're not calling `save_pretrained()` during training automatically, but you can always load your lightning module from a vanilla PL checkpoint, then call `save_pretrained()` directly. This is more explicit and easier to understand rather than trying to be smart with callbacks.

In my case, I have a BertForSequenceClassification obj within the PL module:

```python
  class LMForSequenceClassification(pl.LightningModule):
      def __init__(
              self,
              src_model: str,
              learning_rate: float
          ):
          super().__init__()
          self.save_hyperparameters()
          self.model = BertForSequenceClassification.from_pretrained(src_model)

      def forward(self, **inputs):
          return self.model(**inputs, output_attentions=True, return_dict=True)

      def training_step(self, batch, batch_idx):
          output = self(**batch)
          loss = output["loss"]
          self.log('train_loss', loss)
          return loss

      def validation_step(self, batch, batch_idx):
          output = self(**batch)
          loss = output["loss"]
          self.log('val_loss', loss)

      def test_step(self, batch, batch_idx):
          output = self(**batch)
          loss = output["loss"]
          self.log("test_loss", loss)

      def configure_optimizers(self):
          optimizer = torch.optim.Adam(
              self.parameters(), lr=self.hparams.learning_rate
          )
          return optimizer

      def get_backbone(self):
          return self.model
``` 

and once my training has completed as here: 

```python
model = LMForSequenceClassification(src_model, learning_rate)

# set some training stuff (logger, model ckpt, early stopping)
mlf_logger = pl.loggers.MLFlowLogger(
    experiment_name=experiment_name,
    tracking_uri="file:./mlruns"
)
aim_logger = AimLogger(experiment=experiment_name)
early_stopping = pl.callbacks.EarlyStopping(
    "val_loss",
    patience=early_stop_epochs
)

model_dir = os.path.join(output_dir, f"vanilla-wiki-{seed}")
model_checkpoint = pl.callbacks.ModelCheckpoint(
    monitor="val_loss",
    dirpath=model_dir,
    filename="{epoch}-{val_loss:.3f}-{train_loss:.3f}",
)

trainer = pl.Trainer(
    gpus=gpus,
    max_epochs=max_epochs,
    logger=[mlf_logger, aim_logger],
    callbacks=[early_stopping, model_checkpoint]
)
trainer.fit(model, train_loader, val_loader)

# test on the dataset in-distribution
trainer.test(test_dataloaders=test_loader, ckpt_path="best")

# Save the tokenizer and the backbone LM with HuggingFace's serialization.
# To avoid mixing PL's and HuggingFace's serialization:
# https://github.com/PyTorchLightning/pytorch-lightning/issues/3096#issuecomment-686877242
best_PL = LMForSequenceClassification.load_from_checkpoint(
    model_checkpoint.best_model_path
)
best_PL.get_backbone().save_pretrained(model_dir)
tokenizer.save_pretrained(model_dir)
```

I'd like to load the best checkpoint and save its BertForSequenceClassification backbone with HuggingFace's `save_pretrained` (that is a requirement) (last lines in the code above).

However, where I load the PL class from the checkpoint, I suppose the weights and biases of the Bert's classification head can't be loaded for some sort of difference in the naming convention used by PL during the serialization. The log from which I can suppose that is:

```
GPU available: True, used: True
INFO:2021-01-27 11:15:50,686:distributed:GPU available: True, used: True
TPU available: None, using: 0 TPU cores
INFO:2021-01-27 11:15:50,686:distributed:TPU available: None, using: 0 TPU cores
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:2021-01-27 11:15:50,686:accelerator_connector:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name  | Type                          | Params
--------------------------------------------------------
0 | model | BertForSequenceClassification | 109 M
--------------------------------------------------------
109 M     Trainable params
0         Non-trainable params
109 M     Total params
INFO:2021-01-27 11:15:52,484:lightning:
  | Name  | Type                          | Params
--------------------------------------------------------
0 | model | BertForSequenceClassification | 109 M
--------------------------------------------------------
109 M     Trainable params
0         Non-trainable params
109 M     Total params
Epoch 0: 100%|█████████████████████████████████████████████████| 999/999 [01:40<00:00,  9.94it/s, loss=0.107, v_num=f5cd]
Testing: 100%|█████████████████████████████████████████████████████████████████████████| 249/249 [00:09<00:00, 25.20it/s]
--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'test_loss': tensor(0.1145, device='cuda:0')}
--------------------------------------------------------------------------------
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
``` 

Specifically, from the last lines of the log, I can see that the classifier weights and biases have not been loaded, which is potentially harmful in my case as I've just fine-tuned them from my downstream task.

Am I missing something? Should I go for the custom-checkpoint saving callback, somewhat related to what @rohitgr7 suggested for the trainer?

Thanks!

**Comment by junwen-austin** [comment]
The warning message you saw from last line is actually fine because at test, PL will initialize your model again from the pre-trained BERT model, and then load your fine-tuned model. The warning message you saw is for the loading of the pre-trained BERT model and those "un-trained" weights will be overridden by those from your fine-tuned model.

**Comment by g8a9** [comment]
I don't think that is the case. When I restore the best model from my ModelCheckpoint, the initialization of the parameters `['classifier.weight', 'classifier.bias']` fails. In fact, the parameters
`['cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias']` stored in the checkpoint are not used.

Also, I feel like I'm exactly in this case from the warning message:
*This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).*

**Comment by junwen-austin** [comment]
what do you mean by the "initialization of the parameters ['classifier.weight', 'classifier.bias'] fails"? 

You should just receive the warning during testing.

**Comment by g8a9** [comment]
Ok, *fail* is not the most appropriate verb. 
The parameters `classifier.weight` and `classifier.bias` are being randomly initialized when I load the checkpoint, **which shouldn't be the case as I just fine-tuned the model and saved it with all the parameters**.

(I've also checked them, they are different).

**Comment by junwen-austin** [comment]
I do not think you understand what I said above, unfortunately. Let me try this way. The loading of PL model happens in two stages:

1.  call `__init__`: this is the place a pre-trained BERT model is loaded and weights and bias of classifier are randomly initialized and therefore the warning appears
2. PL load the fine-tuned model to overwrite weights and bias of pre-trained BERT model

What you can do to verify the fine-tuned model is actually loaded is that you could extract the pure BERT model after it is loaded:
```
model = LMForSequenceClassification.load_from_checkpoint(...)
model.model.save_pretrained('your new path for the pure fine-tuned BERT model without any PL')
```

then you could load it with Transformers
```
fine_tuned_model = BertForSequenceModel.from_pretrained('your new path for the pure fine-tuned BERT model without any PL')
```

Hope this helps.

**Comment by g8a9** [comment]
Ok, I didn't get the two-step initialization.
I'll give your example a try. Thank you for your help!

**Comment by BuxianChen** [comment]
```python
class HfModelCheckpoint(ModelCheckpoint):
    def _save_checkpoint(self, trainer: "pl.Trainer", filepath: str) -> None:
        super()._save_checkpoint(trainer, filepath)
        hf_save_dir = filepath+".dir"
        if trainer.is_global_zero:
            # 
            trainer.lightning_module.model.save_pretrained(hf_save_dir)
            trainer.lightning_module.tokenizer.save_pretrained(hf_save_dir)
    
    # https://github.com/Lightning-AI/lightning/pull/16067
    def _remove_checkpoint(self, trainer: "pl.Trainer", filepath: str) -> None:
        super()._remove_checkpoint(trainer, filepath)
        hf_save_dir = filepath+".dir"
        if trainer.is_global_zero:
            fs, _ = url_to_fs(hf_save_dir)
            if fs.exists(hf_save_dir):
                fs.rm(hf_save_dir, recursive=True)
```

**Comment by yuto3o** [comment]
if I only want to save trainable parameters in the checkpoint. Do anyone have any idea?

**Comment by xooca** [comment]
I think if you have customized layers in the forward function then it's better to use PL method of prediction.
